service/matching/version_graph.go

```diff
@@ -0,0 +1,223 @@
+			}
+		} else {
+			// Check if the version is already a default, and remove it from being one if it is.
+			curDefault := existingData.GetCurrentDefault()
+			if curDefault.GetVersion().Equal(req.GetVersionId()) {
+				existingData.CurrentDefault = nil
+				if curDefault.GetPreviousCompatible() != nil {
+					existingData.CurrentDefault = curDefault.GetPreviousCompatible()
+				} else if curDefault.GetPreviousIncompatible() != nil {
+					existingData.CurrentDefault = curDefault.GetPreviousIncompatible()
+				}
+				return nil
+			}
+			// Check if it's a compatible leaf, and remove it from being one if it is.
+			for i, def := range existingData.GetCompatibleLeaves() {
+				if def.GetVersion().Equal(req.GetVersionId()) {
+					existingData.CompatibleLeaves =
+						append(existingData.CompatibleLeaves[:i], existingData.CompatibleLeaves[i+1:]...)
+					if def.GetPreviousCompatible() != nil {
+						existingData.CompatibleLeaves =
+							append(existingData.CompatibleLeaves, def.GetPreviousCompatible())
+					}
+					return nil
+				}
+			}
+			return serviceerror.NewInvalidArgument(
+				"requests to update build id ordering cannot create a new non-default version with no links")
+		}
+	}
+	return nil
+}
+
+// Finds the node that the provided version should point at, given that it says it's compatible with the provided
+// version. Note that this does not necessary mean *that* node. If the version being targeted as compatible has nodes
+// which already point at it as their previous compatible version, that chain will be followed out to the leaf, which
+// will be returned.
+func findCompatibleNode(
+	existingData *persistence.VersioningData,
+	versionId *taskqueuepb.VersionId,
+) (*taskqueuepb.VersionIdNode, int) {
+	// First search down from all existing compatible leaves, as if any of those chains point at the desired version,
+	// we will need to return that leaf.
+	for ix, node := range existingData.GetCompatibleLeaves() {
+		if node.GetVersion().Equal(versionId) {
+			return node, ix
+		}
+		if findInNode(node, versionId) != nil {
+			return node, ix
+		}
+	}
+	// Otherwise, this must be targeting some version in the default/incompatible chain, and it will become a new leaf
+	curDefault := existingData.GetCurrentDefault()
+	// There can only be one version of each type in the default list, so ignore any others.
+	if reflect.TypeOf(versionId) == reflect.TypeOf(curDefault.GetVersion()) {
+		if curDefault.GetVersion().Equal(versionId) {
+			return curDefault, -1
+		}
+		if nn := findInNode(curDefault, versionId); nn != nil {
+			return nn, -1
+		}
+	}
+
+	return nil, -1
+}
+
+func findInNode(
+	node *taskqueuepb.VersionIdNode,
+	versionId *taskqueuepb.VersionId,
+) *taskqueuepb.VersionIdNode {
+	if node.GetVersion().Equal(versionId) {
+		return node
+	}
+	if node.GetPreviousCompatible() != nil {
+		return findInNode(node.GetPreviousCompatible(), versionId)
```

Comment:
I hate to say this, but I think we should use an explicit loop to be safe. the go compiler is not required to do tce

Reply:
Hmm. Will that be a worry since the graph is gonna be constrained in size anyway?

Reply:
where is the constraint happening?

Reply:
Nowhere yet. Need to add it. Initially I thought I'd do it in another PR, but, probably should go in this one too.

Reply:
Added